lib/services/extensions-service-base.ts

```diff
@@ -0,0 +1,99 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+import path = require("path");
+import options = require("../common/options");
+import helpers = require("../helpers");
+import temp = require("temp");
+temp.track();
+
+export class ExtensionsServiceBase {
+	private extensionVersions: IStringDictionary = {};
+
+	constructor(public $fs: IFileSystem,
+				public $httpClient: Server.IHttpClient,
+				public $logger: ILogger) {
+		if (this.$fs.exists(this.versionsFile).wait()) {
+			this.extensionVersions = this.$fs.readJson(this.versionsFile).wait() || {};
+		}
+	}
+
+	public get cacheDir(): string {
+		return path.join(options.profileDir, "Cache");
+	}
+
+	public getExtensionVersion(packageName: string): string {
+		return this.extensionVersions[packageName];
+	}
+
+	public getExtensionPath(packageName: string): string {
+		return path.join(this.cacheDir, packageName);
+	}
+
+	public prepareExtensionBase(extensionData: IExtensionData, initialCachedVersion: string, beforeDownloadAction?: () => IFuture<void>): IFuture<void> {
+		return ((): void => {
+			if (this.$fs.exists(this.versionsFile).wait()) {
+				this.extensionVersions = this.$fs.readJson(this.versionsFile).wait() || {};
+			}
+
+			var packageName = extensionData.packageName;
+			var extensionVersion = extensionData.version;
+			var extensionPath = extensionData.pathToSave || this.getExtensionPath(extensionData.packageName);
+
+			var cachedVersion = initialCachedVersion;
+			this.$logger.debug("Server version: %s", extensionVersion);
+
+			if (this.extensionVersions[packageName]) {
+				cachedVersion = this.extensionVersions[packageName];
+				this.$logger.debug("Cached version is: %s", cachedVersion);
+			}
+
+			if( helpers.versionCompare(cachedVersion, extensionVersion) < 0) {
+				this.$logger.info("Updating %s package...", packageName);
+				var zipFileName = temp.path({ path:  path.join(this.cacheDir, packageName + ".zip") });
+
+				if(beforeDownloadAction) {
+					beforeDownloadAction().wait();
+				}
+
+				this.$fs.deleteDirectory(extensionPath).wait();
+
+				this.$fs.createDirectory(extensionPath).wait();
+				this.$logger.trace("Extension path for %s: %s", packageName, extensionPath);
+
+				try {
+					this.downloadPackage(extensionData.downloadUri, zipFileName).wait();
+					this.$fs.unzip(zipFileName, extensionPath).wait();
+					this.$fs.deleteFile(zipFileName).wait();
+					this.extensionVersions[packageName] = extensionVersion;
+					this.saveVersionsFile().wait();
+				} catch(err) {
+					this.$fs.deleteDirectory(extensionPath).wait();
+					throw err;
+				}
+				this.$logger.info("Finished updating %s package.", packageName);
```

Comment:
There's a significant delay after printing this message on the console under linux. I suggest we add some busy indicator to tell the user that we are working.